### PR TITLE
cache by hand

### LINKED ISSUE #2131: duplicated method for `primary_decomposition`
**Describe the bug**

When loading Oscar (on 3d098c78ccf9ee05932baa3e8c97059cb2373b8e), I get
```
WARNING: Method definition primary_decomposition(T) where {T<:(Oscar.MPolyIdeal{S} where S)} in module Oscar at /Users/bla/.julia/dev/Oscar/src/Rings/mpoly-ideals.jl:498 overwritten at /Users/bla/.julia/dev/Oscar/src/Rings/mpoly-ideals.jl:517.
```

**To Reproduce**

```
using Oscar
```

I suspect #2122. @simonbrandhorst

### PR DESCRIPTION
**Pull request by simonbrandhorst**
Resolves  #2131

### CODE REVIEW

**Review by thofma — APPROVED** [review]

File: src/Rings/mpoly-ideals.jl [hunk 1]
```diff
@@ -495,27 +495,27 @@ julia> L = primary_decomposition(I)
...
 """
-function primary_decomposition(I::MPolyIdeal; alg=:GTZ)
-  R = base_ring(I)
-  singular_assure(I)
-  if elem_type(base_ring(R)) <: FieldElement
-    if alg == :GTZ
-      L = Singular.LibPrimdec.primdecGTZ(I.gens.Sx, I.gens.S)
-    elseif alg == :SY
-      L = Singular.LibPrimdec.primdecSY(I.gens.Sx, I.gens.S)
+function primary_decomposition(I::T, alg=:GTZ) where {T<:MPolyIdeal}
```
Comment (thofma) [inline src/Rings/mpoly-ideals.jl]:
```suggestion
function primary_decomposition(I::T; alg=:GTZ) where {T<:MPolyIdeal}
```

**Comment by thofma** [comment]
The examples in the docstring are not working anymore:
```
│ julia> L = primary_decomposition(I)
│ 3-element Vector{Tuple{MPolyIdeal{QQMPolyRingElem}, MPolyIdeal{QQMPolyRingElem}}}:
│  (ideal(x^3 - x - y^2), ideal(x^3 - x - y^2))
│  (ideal(x^2 - 2*x*y - 2*x + y^2 + 2*y + 1), ideal(x - y - 1))
│  (ideal(y, x^2), ideal(x, y))
│ 
│ julia> L = primary_decomposition(I, alg = :SY)
│ 3-element Vector{Tuple{MPolyIdeal{QQMPolyRingElem}, MPolyIdeal{QQMPolyRingElem}}}:
│  (ideal(x^3 - x - y^2), ideal(x^3 - x - y^2))
│  (ideal(x^2 - 2*x*y - 2*x + y^2 + 2*y + 1), ideal(x - y - 1))
│  (ideal(y, x^2), ideal(y, x))
```

Because of caching, the second call does not do anything at all and just returns the same as without `:alg` argument.

How to proceed @simonbrandhorst?

**Comment by simonbrandhorst** [comment]
We provide an optional argument `cache=true`.

**Comment by simonbrandhorst** [comment]
Alternatively one can give a keyword `recompute=false`

**Comment by thofma** [comment]
Will `cache = false` also not store the result or just compute it fresh?

**Comment by wdecker** [comment]
@thofma @simonbrandhorst :  Typically, we would call `primary_decomposition(I, alg = :SY)` only if `primary_decomposition(I)` does not succeed (that is, no primary decompositin has been computed yet). Thus, I suggest to leave things in the code how they are, but remove

`julia> L = primary_decomposition(I, alg = :SY)`

etc. from the docstring.

**Comment by thofma** [comment]
@wdecker that sounds like a reasonable solution.

**Comment by simonbrandhorst** [comment]
I would bypass the cache completely.

**Pull request merged**